lib/std/math/powi.zig

```diff
@@ -10,108 +10,92 @@ const testing = std.testing;
-            //  powi(x, y)     = Underflow for for y > @sizeOf(x) - 1 y < 0
-            const bit_size = @sizeOf(T) * 8;
-            if (info.Int.signedness == .signed) {
-                if (x == -1) {
-                    //  powi(-1, y)    = -1 for for y an odd integer
-                    //  powi(-1, y)    = 1 for for y an even integer
-                    if (@mod(y, 2) == 0) {
-                        return 1;
-                    } else {
-                        return -1;
-                    }
-                }
-
-                if (x > 0 and y >= bit_size - 1) {
-                    return error.Overflow;
-                } else if (x < 0 and y > bit_size - 1) {
-                    return error.Underflow;
-                }
-            } else {
-                if (y >= bit_size) {
-                    return error.Overflow;
-                }
-            }
+    if (x == -1) {
+        return -1;
+    }
 
-            var base = x;
-            var exp = y;
-            var acc: T = 1;
-
-            while (exp > 1) {
-                if (exp & 1 == 1) {
-                    if (@mulWithOverflow(T, acc, base, &acc)) {
-                        if (x > 0) {
-                            return error.Overflow;
-                        } else {
-                            return error.Underflow;
-                        }
-                    }
-                }
-
-                exp >>= 1;
-
-                if (@mulWithOverflow(T, base, base, &base)) {
-                    if (x > 0) {
-                        return error.Overflow;
-                    } else {
-                        return error.Underflow;
-                    }
-                }
-            }
+    if (x == 0) {
+        if (y > 0) {
+            return 0;
+        } else {
+            // not overflow in strict sense
+            return error.Overflow;
+        }
+    }
+    // x >= 2 or x <= -2
+    if (y >= bit_size) {
+        return error.Overflow;
+    }
+    if (y < 0) {
+        return error.Underflow;
+    }
 
-            if (exp == 1) {
-                if (@mulWithOverflow(T, acc, base, &acc)) {
-                    if (x > 0) {
-                        return error.Overflow;
-                    } else {
-                        return error.Underflow;
-                    }
-                }
+    var base = x;
+    var exp = y;
+    var acc: T = if (does_one_overflow) unreachable else 1;
+
+    while (exp > 1) {
+        if (exp & 1 == 1) {
+            if (@mulWithOverflow(T, acc, base, &acc)) {
+                return error.Overflow;
             }
+        }
+
+        exp >>= 1;
+
+        if (@mulWithOverflow(T, base, base, &base)) {
+            return error.Overflow;
+        }
+    }
 
-            return acc;
-        },
+    if (exp == 1) {
+        if (@mulWithOverflow(T, acc, base, &acc)) {
+            return error.Overflow;
+        }
     }
+
+    return acc;
 }
 
 test "math.powi" {
```

Comment:
Unfortunately this is missing a bunch of edge cases.
Using pseudo code related to python
```python
0**0 == 1 // u0,i0,u1,i1
1**1 == 1 // same
1**0 == 1
1**y == 1, y any
```

Reply:
it is on the first branch, `if (x == 1 or y == 0`

Reply:
I dont understand. The special cases `u0,u1,i0,i1`, which you added in the PR, are not tested.

Reply:
i misunderstood your comment, will add asap

Reply:
sorry, i misunderstood your comment. 
now i added test cases on `u0, i0, i1`. `u1` is not special cases as it can represent `1`.